README.md
# Coloriss 

[![Build Status](https://travis-ci.org/tbzr/coloriss.png)](https://travis-ci.org/tbzr/coloriss)
[![npm version](https://badge.fury.io/js/coloriss.png)](https://badge.fury.io/js/coloriss)

RGB | Hex converter. Easy to implement for both client-side or server-side services.

## Installation
You can install it manually or using some dependencies managers.
Using npm:
```
npm install --save coloriss
```
Using bower:
```
bower install --save coloriss
```

## Usage
Using in a node.js script
```javascript
var Coloriss = require('coloriss');

var cs = new Coloriss('#fff');

console.log(cs.hex()); // output: "#FFFFFF"
console.log(cs.rgb()); // output: {r:255, g:255, b:255}
console.log(cs.rgb(true)); // output: "rgb(255, 255, 255)"
```

Or directly in a javascript
```javascript
<script>

	var Coloriss = new Coloriss('#fff');
	...
	cs.hex();

</script>
```

## Demo
See a little demo of Coloriss implementation here: [Demo Page](http://tbzr.github.io/coloriss)

## License
Coloriss is 


CHANGELOG.md

## 1.1.0
- Introduce rgbAsStr() / rgbaAsStr() methods
- Introduce boolean parameter to transform as a string rgb/rgba value with rgb()/rgba() methods
- Introduce colors() method that give you an object of pretty cool colors
- Introduce randomColor() method that pick up a random color from the colors object
- Fixing some conversion bugs

test.js
var assert = require('assert');
var Coloriss = require('./coloriss');

describe("01 - Init", function () {

	// Init

});

describe("02 - Hex colors", function () {
	
	// hex() func called

	it("should convert Hex color into valid 32bits hexadecimal color - 1", function () {
		var cs = new Coloriss('fff');
		assert.strictEqual(cs.hex(), '#FFFFFF');
	});

	it("should convert Hex color into valid 32bits hexadecimal color - 2", function () {
		var cs = new Coloriss('#000');
		assert.strictEqual(cs.hex(), '#000000');
	});

	it("should convert Hex color into valid 32bits hexadecimal color - 3", function () {
		var cs = new Coloriss('#FFFFFF');
		assert.strictEqual(cs.hex(), '#FFFFFF');
	});

	it("should convert Hex color into valid 32bits hexadecimal color - 4", function () {
		var cs = new Coloriss('#FfF23F');
		assert.strictEqual(cs.hex(), '#FFF23F');
	});

	it("should convert RGB color into valid 32bits hexadecimal color - 1", function () {
		var cs = new Coloriss('rgb(255, 255, 255)');
		assert.strictEqual(cs.hex(), '#FFFFFF');
	});

	it("should convert RGBA color into valid 32bits hexadecimal color - 1", function () {
		var cs = new Coloriss('rgba(255, 255, 255, 1)');
		assert.strictEqual(cs.hex(), '#FFFFFF');
	});

});

describe("03 - RGB colors", function () {

	// rgb() func called

	it("should convert hexadecimal color into RGB via color() method - 1", function () {

		var cs, rgb, ret;

		cs = new Coloriss();
		ret = cs.color('#fff').rgb();
		rgb = {r:255, g:255, b:255};
		assert.equal(typeof ret, 'object');
		assert.deepEqual(ret, rgb);
	});

	it("should convert hexadecimal color into valid RGB color - 1", function () {

		var cs, rgb, ret;

		cs = new Coloriss('#FfFfFF');
		ret = cs.rgb();
		rgb = {r:255, g:255, b:255};
		assert.equal(typeof ret, 'object');
		assert.deepEqual(ret, rgb);
	});

	it("should convert hexadecimal color into valid RGB color - 2", function () {

		var cs, rgb, ret;

		cs = new Coloriss('#000');
		ret = cs.rgb();
		rgb = {r:0, g:0, b:0};
		assert.equal(typeof ret, 'object');
		assert.deepEqual(ret, rgb);
	});

	it("should convert RGB color into valid RGB color - 1", function () {

		var cs, rgb, ret;

		cs = new Coloriss('rgb(0, 0, 0)');
		ret = cs.rgb();
		rgb = {r:0, g:0, b:0};
		assert.equal(typeof ret, 'object');
		assert.deepEqual(ret, rgb);
	});

	it("should convert RGBA color into valid RGB color - 1", function () {

		var cs, rgb, ret;

		cs = new Coloriss('rgba(0, 0, 0, 0.3)');
		ret = cs.rgb();
		rgb = {r:0, g:0, b:0};
		assert.equal(typeof ret, 'object');
		assert.deepEqual(ret, rgb);
	});

	it("should convert RGB color into valid RGB color as String - 1", function () {

		var cs, rgb;

		rgb = "rgb(0, 0, 5)";
		cs = new Coloriss(rgb);
		assert.strictEqual(cs.rgb(true), rgb);
	});

	it("should convert RGB color into valid RGB color as String using dedicated method - 1", function () {

		var cs, rgb;

		rgb = "rgb(0, 0, 5)";
		cs = new Coloriss(rgb);
		assert.strictEqual(cs.rgbAsStr(), rgb);
	});

});

describe('04 - RGBA colors', function () {

	// rgba() func called

	it("should convert hexadecimal color into RGBA via color() method - 1", function () {

		var cs, rgba, ret;

		cs = new Coloriss();
		ret = cs.color('#fff').rgba();
		rgba = {a:1, r:255, g:255, b:255};
		assert.equal(typeof ret, 'object');
		assert.deepEqual(ret, rgba);
	});

	it("should convert RGB color into RGBA color - 1", function () {

		var cs, rgba;

		rgba = {a:1, r:255, g:255, b:255};
		cs = new Coloriss('rgb(255, 255, 255)');
		assert.deepEqual(cs.rgba(), rgba);
	});

	it("should convert hexadecimal color into RGBA color - 1", function () {

		var cs, rgba;

		rgba = {a:1, r:255, g:255, b:255};
		cs = new Coloriss('#fff');
		assert.deepEqual(cs.rgba(), rgba);
	});

	it("should convert RGBA color into RGBA color - 1", function () {

		var cs, rgba;

		rgba = {a:0.42, r:255, g:255, b:255};
		cs = new Coloriss('rgb(255, 255, 255, 0.42)');
		assert.deepEqual(cs.rgba(), rgba);
	});

	it("should convert RGBA color into valid RGBA color as String - 1", function () {

		var cs, rgba;

		rgba = "rgba(10, 210, 115, 0.2)";
		cs = new Coloriss(rgba);
		assert.strictEqual(cs.rgba(true), rgba);
	});

	it("should convert RGBA color into valid RGBA color as String using dedicated method - 1", function () {

		var cs, rgb;

		rgb = "rgba(10, 210, 115, 0.2)";
		cs = new Coloriss(rgb);
		assert.strictEqual(cs.rgbaAsStr(), rgb);
	});	
});

describe('04 - Individual colors', function () {

	it("should convert hexadecimal color into correct red", function () {
		var cs = new Coloriss('#4200ff');
		assert.strictEqual(cs.red(), 66);
	});

	it("should convert hexadecimal color into correct green", function () {
		var cs = new Coloriss('#4200ff');
		assert.strictEqual(cs.green(), 0);
	});

	it("should convert hexadecimal color into correct blue", function () {
		var cs = new Coloriss('#4200ff');
		assert.strictEqual(cs.blue(), 255);
	});

	it("should convert RGBA color into correct alpha", function () {
		var cs = new Coloriss('rgba(255, 255, 255, 0.42)');
		assert.strictEqual(cs.alpha(), 0.42);
	});

	it("should convert hexadecimal color into correct red", function () {
		var cs = new Coloriss('#fff');
		assert.strictEqual(cs.grayScale(), 255);
	});

});


test/test.js

describe("01 - Init", require('./01_init'));
describe("02 - Hex colors", require('./02_hex'));
describe("03 - RGB colors", require('./03_rgb'));
describe('04 - RGBA colors', require('./04_rgba'));
describe('04 - Individual colors', require('./05_colors'));


dev/app.js

angular.module('Coloriss', [])

.controller('AppCtrl', function ($scope) {

	var colors = {
		red: '#C0392B',
		green: '#2ECC71',
		blue: '#3498DB',
		yellow: '#F1C40F',
		orange: '#E67E22',
		turquoise: '#1ABC9C',
		grey: '#EEEEEE',
		black: '#000000',
		dark: '#333333',
		white: '#FFFFFF',

		// New colors 1.1
		pink: '#F7CAC9',
		gold: '#FFD700',
		silver: '#C0C0C0',
		copper: '#B87333',
		brass: '#B5A642',
		azur: '#007FFF',
		serenity: '#92A8D1'

	};

	$scope.form = {};
	$scope.invert = false;

	$scope.needInvert = function () {
		var color = angular.copy($scope.colors.rgb);
		var gs = Math.floor((color.r + color.g + color.b) / 3);
		return (gs >= 100);
	};

	$scope.reinit = function () {

		$scope.colors = {
			hex: '000000',
			rgb: {r: -1, g: -1, b: -1}
		};
	};

	$scope.listen = function (event) {
		if (event.keyCode === 27 && event.which === 27) {
			$scope.form.color = '';
			$scope.reinit();
			$scope.invert = false;
		}
	};

	$scope.update = function () {

		var color, cs, i;
		
		color = angular.copy($scope.form.color);

		if (!$scope.form.color)
			$scope.reinit();

		for (i in colors) {

			if (color.toLowerCase() === i) {
				color = colors[i];
			}
		}

		if (color.length > 3) {
			// Call Coloriss API.
			cs = new Coloriss(color);
			$scope.colors.hex = cs.hex();
			$scope.colors.rgb = cs.rgb();
		}

		$scope.invert = $scope.needInvert() ? true : false;
	};

	$scope.reinit();

})


dev-02082016.js

/**
 * This is the main Coloriss core file
 * - Everything still under development. But you can test it at https://tbzr.github.io/coloriss
 */

(function () {

	var root = typeof self == 'object' && self.self === self && self ||
            typeof global == 'object' && global.global === global && global ||
            this;

    function Coloriss (color) {
    	if (typeof color !== 'undefined') {
    		parse(color);
    	}
    	return this;
    };

    if (typeof exports != 'undefined' && !exports.nodeType) {
    	if (typeof module != 'undefined' && !module.nodeType && module.exports) {
    		exports = module.exports = Coloriss;
    	}
    	exports.Coloriss = Coloriss;
    } else {
    	root.Coloriss = Coloriss;
    }

    var intToHex = function (n) {
    	var s = "0123456789ABCDEF";

    	n = parseInt(n, 10);
    	if (!isNaN(n)) {
    		n = Math.max(0, Math.min(n, 255));
    		return s.charAt((n - n % 16) / 16) + s.charAt(n % 16);
    	}
    };

    var hexToRGB = function (hex) {
    	var a, r, g, b;

    	if (hex.indexOf('#') > -1)
    		hex = hex.substr(1, hex.length);

    	if (hex.length === 3) {
    		hex += hex;
    	}

    	hex = parseInt("0x" + hex.toUpperCase());

		a = hex.length === 6 ? '1' : hex >> 24;
		r = 0xFF & (hex >> 16);
		g = 0xFF & (hex >> 8);
		b = 0xFF & hex;

		return {a:a, r:r, g:g, b:b};
    };

    var RGBtoHex = function (rgb) {

    	if (typeof rgb === 'object') {
    		return intToHex(rgb.r) + intToHex(rgb.g) + intToHex(rgb.b);
    	}
    };

    var parse = function (color) {

    	var match;

    	// Match an Hexadecimal color
    	if ((match = color.match(/^#?(?:[0-9a-f]{3}){1,2}$/i))) {
    		this._hex = match[0];
    		this._rgb = hexToRGB(this._hex);
    	}
    	// Match an rgb(a) color
    	else if ((match = color.match(/^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*(\d+(?:\.\d+)?))?\)$/))) {
    		match.shift();
    		this._rgb = {a:match[3], r:match[0], g:match[1], b:match[2]};
    		this._hex = RGBtoHex(this._rgb);
    	}
    };

    Coloriss.VERSION = '1.0.0';

    Coloriss.COLORS = {
    	red: '#C0392B',
    	green: '#2ECC71',
    	blue: '#3498DB',
    	yellow: '#F1C40F',
    	orange: '#E67E22',
    	turquoise: '#1ABC9C',
    	grey: '#EEEEEE',
    	black: '#000000',
    	dark: '#333333',
    	white: '#FFFFFF'
    };

    Coloriss.prototype.hex = function () {
    	return this._hex;
    };

    Coloriss.prototype.rgb = function () {
    	return this._rgb;
    };

    Coloriss.prototype.alpha = function () {
    	return this._rgb.a;
    };

    Coloriss.prototype.red = function () {
    	return this._rgb.r;
    };

    Coloriss.prototype.green = function () {
    	return this._rgb.g;
    };

    Coloriss.prototype.blue = function () {
    	return this._rgb.b;
    };

    Coloriss.prototype.greyScale = function () {
    	var c = this._rgb;
    	return (c.r + c.g + c.b) / 3;
    };

})();

coloriss.js

(function (root) { 'use strict';

    var Coloriss = (function () {

        var _colors;

        _colors = {
            red: '#C0392B',
            green: '#2ECC71',
            blue: '#3498DB',
            yellow: '#F1C40F',
            orange: '#E67E22',
            turquoise: '#1ABC9C',
            grey: '#EEEEEE',
            black: '#000000',
            dark: '#333333',
            white: '#FFFFFF',
            pink: '#F7CAC9',
            gold: '#FFD700',
            silver: '#C0C0C0',
            copper: '#B87333',
            brass: '#B5A642',
            azur: '#007FFF',
            serenity: '#92A8D1'
        };

        var Coloriss = function (color, options) {
            this._hex = '';
	    this._rgb = {};
            // Options not used yet.
            if (typeof color !== 'undefined') {
                this.parse(color);
            }
        };

        var intToHex = function (n) {
            var s = "0123456789ABCDEF";

            n = parseInt(n, 10);
            if (!isNaN(n)) {
                n = Math.max(0, Math.min(n, 255));
                return s.charAt((n - n % 16) / 16) + s.charAt(n % 16);
            }
        };

        var sanitizeHex = function (hex) {

		if (hex.indexOf('#') > -1)
			hex = hex.substr(1, hex.length);
		if (hex.length === 3)
			hex += hex;

		return hex.toUpperCase();
        };

        var hexToRGB = function (hex) {
            var a, r, g, b;

            hex = parseInt("0x" + sanitizeHex(hex));

            a = 1;
            r = 0xFF & (hex >> 16);
            g = 0xFF & (hex >> 8);
            b = 0xFF & hex;

            return {r:r, g:g, b:b, a:a};
        };

        var RGBtoHex = function (rgb) {

            if (typeof rgb === 'object') {
                return intToHex(rgb.r) + intToHex(rgb.g) + intToHex(rgb.b);
            }
        };

        var toString = function (obj) {
            var s, n, v = [];

            s = 'rgb';
            s += (typeof obj.a !== 'undefined') ? 'a(' : '(';
            Object.keys(obj).map(function (key, i) {
                v.push(obj[key]);
            });
            s += v.join(', ') + ')';
            return s;
        };

        /**
         * Main function, determine which type of color the value passed 
         * in parameter is. And then format as rgb/hex
         */ 
        Coloriss.prototype.parse = function (color) {

            var match;

            // Match an Hexadecimal color
            if ((match = color.match(/^#?(?:[0-9a-f]{3}){1,2}$/i))) {
                this._hex = '#' + sanitizeHex(match[0]);
                this._rgb = hexToRGB(this._hex);
            }
            // Match an rgb(a) color
            else if ((match = color.match(/^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*(\d+(?:\.\d+)?))?\)$/))) {
                match.shift();
                this._rgb = {
                    r: parseInt(match[0]),
                    g: parseInt(match[1]),
                    b: parseInt(match[2]),
                    a: parseFloat(match[3])
                };
                this._hex = '#' + RGBtoHex(this._rgb);
            }
        };

        /**
         * Same as the constructor but without options.
         */
        Coloriss.prototype.color = function (c) {
        	this.parse(c);
        	return this;
        };

        Coloriss.prototype.hex = function () {
            return this._hex;
        };

        Coloriss.prototype.rgb = function (str) {
            delete this._rgb.a;
            if (typeof str !== 'undefined' && str) {
                return toString(this._rgb);
            }
            return this._rgb;
        };

        Coloriss.prototype.rgba = function (str) {
        	if (!this._rgb.a || typeof this._rgb.a === 'undefined')
        		this._rgb.a = 1;
            if (typeof str !== 'undefined' && str) {
                return toString(this._rgb);
            }
            return this._rgb;
        };

        Coloriss.prototype.rgbAsStr = function () {
            delete this._rgb.a;
            return toString(this._rgb);
        };

        Coloriss.prototype.rgbaAsStr = function () {
            return toString(this._rgb);
        };

        Coloriss.prototype.alpha = function () {
            return this._rgb.a;
        };

        Coloriss.prototype.red = function () {
            return this._rgb.r;
        };

        Coloriss.prototype.green = function () {
            return this._rgb.g;
        };

        Coloriss.prototype.blue = function () {
            return this._rgb.b;
        };

        Coloriss.prototype.grayScale = function () {
            var c = this._rgb;
            return (c.r + c.g + c.b) / 3;
        };

        Coloriss.prototype.colors = function () {
            return this._colors;
        };

        /**
         * Pickup a random color from the _colors object
         */
        Coloriss.prototype.randomColor = function () {

            var res, i, c = 0;

            for (i in _colors) {
                if (Math.random() < 1 / ++c)
                    res = i;
            }
            return res;
        };

        return Coloriss;

    })();

    // Export for node.js
    if (typeof exports !== 'undefined') {
        if (typeof module !== 'undefined' && module.exports) {
            module.exports = Coloriss;
        }
        exports.Coloriss = Coloriss;
    }
    // Export for AMD
    else if (typeof define === 'function' && define.amd) {
        define([], function () {
            return Coloriss;
        });
    } else {
        window.Coloriss = Coloriss; 
    }

}).call(this);
